## STAT 207 Lab 3: Sampling, Loops and Functions - [30 total points]

Due: Wednesday, February 17, 11:59pm CST

### SOLUTIONS

This lab consists of 3 main parts. You may add as many cells as you need in order to do the exercises. To add cells use the "Insert" tab from the menu bar above and select "insert cell above" or "insert cell below". 

The data file "cards.csv" is included in the release. It is a data frame representing the 52 cards in a standard deck of playing cards. Each row represents one of the 52 possible cards. For further details see the course notes "04_Sampling".

**Note: Some problems require deriving answers mathematically or explaining rather than coding.**

In [1]:
import pandas as pd

## Question 1: [2 pts] *Calculating* the probability of drawing a five-card hand with two diamonds.

### 1a.  How many combinations of two diamonds can you make from a standard deck of cards?

Hint: How many diamonds are in the deck?

There are 13 diamonds in a standard deck of cards, and if we are choosing two diamonds from these 13 cards, we can represent this as (13 choose 2).

(13 choose 2) = (13!)/((2!)(13-2)!) = 78 ways to choose two diamonds from a standard deck of cards.

### 1b. [2 pts] How many combinations of three non-diamonds can you make from a standard deck of cards?

Hint: How many non-diamonds are in the deck?

There are (52 - 13) = 39 non-diamond cards in the deck. Therefore, there are (39 choose 3) combinations of three non-diamonds cards.
(39 choose 3) = (39!)/((3!)(39-3)!) = 9139 combinations of 3 non-diamond cards. 

### 1c. [2pts]  How many combinations of five card hands contain two diamonds and three non-diamonds?

We can calculate the number of five card hands containing two diamonds and three non-diamonds by this expression:

(number of combinations of two diamonds)(number of combinations of three non diamonds) 
(13 choose 2)(39 choose 3)
(78)(9139)
= 712842 combinations of five card hands that contains two diamonds and three non-diamonds

### 1d. [2 pts] What is the probability of randomly drawing five cards from the deck and getting a hand that has two diamond and three non-diamonds?

We can solve this by taking the number of possible combinations of two diamonds and three non-diamonds and divide that number by the total number of possible 5-carded hands. 

712842/(52 choose 5) = 0.2743 
= 27.43% chance that when randomly drawing 5 cards from a deck, you will pull out two diamonds and three non-diamonds.

## Question 2: *Approximating* the probability of drawing a five-card hand with exactly two diamonds.

### 2a. [1 pt] Read the cards.csv into a dataframe called cards and display the first five rows of the dataframe.

In [5]:
cards = pd.read_csv('cards.csv');
cards.head()

color  suit face
0  black  club    A
1  black  club    2
2  black  club    3
3  black  club    4
4  black  club    5

### 2b. [2 pts] Create a new dataframe called 'sample1' that is a random sample of five rows from this dataframe in the following way.
* Sample without replacemnt.
* Use a random state of 7.

#### Then display this dataframe.

In [43]:
sample1 = cards.sample(5, random_state = 7)
sample1

color     suit face
22  black    spade   10
27    red  diamond    2
34    red  diamond    9
15  black    spade    3
18  black    spade    6

### 2c.  [2 pts] Suppose everyone in the class writes the same code above as you. Would you expect everyone in the class to get the same output as you? Why or why not?

No, because this code is randomly selecting rows from the larger dataframe, so it is extremely unlikely that everyone who runs my code will receive the same data in 'sample1'.

### 2d.  [2 pts] Run this code below to create the series that results from using the .value_counts() function on the 'suits' column of our sample1 dataframe. Then in the code below it, extract just the list of indices for this series.

(Hint: Check Unit 2 slides).

In [44]:
sample1['suit'].value_counts()

spade      3
diamond    2
Name: suit, dtype: int64

In [45]:
sample1.index

Int64Index([22, 27, 34, 15, 18], dtype='int64')

### Tutorial:

We can extract the **value** that corresponds to a named index from a series by using the **.loc()** function.

In [46]:
sample1['suit'].value_counts().loc['spade']

3

### 2e. [1 pt] Use the .loc() function to extract the number of diamonds in sample1.

In [47]:
sample1['suit'].value_counts().loc['diamond']

2

### Tutorial:

We can test whether a certain object is **not in** a list by using the **if**, **else**, and **not in** operators as shown below. 

In [48]:
#Run this
fun_list=['Nick', 'Kevin', 'Joe']

In [49]:
#Run this
if 'Nick' not in fun_list:
    print('this person is not in the list')
else:
    print('this person is in the list')

this person is in the list


In [50]:
#Run this
if 'Harry' not in fun_list:
    print('this person is not in the list')
else:
    print('this person is in the list')

this person is not in the list


### 2f.  [2 pts] Write a code that tests whether 'diamond' is NOT IN the index of sample1['suit'].value_counts().

*  **if** 'diamond' IS NOT in the index, then the **code** should ***print 'there are no diamonds in the hand'***.
    
*  **else** (aka: if 'diamond' IS in the index), then the **code** should ***print 'there are diamonds in the hand'***.

In [51]:
if 'diamond' not in sample1['suit'].value_counts():
    print('there are no diamonds in the hand')
else:
    print('there are diamonds in the hand')

there are diamonds in the hand


### 2g.  [2 pts] Write a code below that does the following.
*First* it tests whether 'diamond' IS NOT in the index of the <u>sample1['suit'].value_counts()</u> series.

*  **if** 'diamond' IS NOT in the index, then the **code** should <u>print 'there are no diamonds in the hand'</u>.
    
*  **else** (aka: if 'diamond' IS in the sample1['suit'].value_counts() variable), then the code should then test whether the **value** that corresponds to the 'diamond' index in <u>sample1['suit'].value_counts()</u> equal to 2.
    - **if** the 'diamond' value does equal 2, then the **code** should <u>print "there are two diamonds in the hand."</u>
        
    - **else** (aka if 'diamond' value does NOT equal 2), then the **code** should print <u>"there are not two diamonds in the hand."</u>
    

#### Hint: Section 9 code in the Unit 5 notebook may be helpful here.

In [52]:
if 'diamond' not in sample1['suit'].value_counts():
    print('there are no diamonds in the hand')
else:
    if sample1['suit'].value_counts().loc['diamond'] == 2:
        print('there are two diamonds in the hand')
    else:
        print('there are not two diamonds in the hand')

there are two diamonds in the hand


### 2h.  [2 pts] Write a FUNCTION below that does the following.

Create a function that is meant to take a dataframe df as input. The input dataframe should represent a sample dataframe of 5 cards/rows that you have sampled from the cards dataframe.
* The function should return **False** if the sample dataframe has no diamonds.
* The function should return a **True** if the sample dataframe contains diamonds and there are exactly two of them.
* The function should also return **False** if the sample dataframe contains diamonds, but there are not exactly two of them.

#### Hint: Section 9 code in the Unit 5 notebook may be helpful here.

In [53]:
def hastwodiamonds (df):
    if 'diamond' not in df['suit'].value_counts():
        return False
    else:
        if df['suit'].value_counts().loc['diamond'] == 2:
            return True
        else:
            return False

### 2i.  [2 pts] Test whether your sample1 dataframe contains exactly two diamonds.

In [54]:
hastwodiamonds(sample1)

True

### 2j.  [2 pts] Use this function to do the following.
1. Simulate drawing 10000 hands (remember hand = 5 cards) from the 52 card deck. Each trial is randomly sampled without replacements.
2. Test whether the trial has exactly two diamonds.
3. Calculates the proportion of the 10000 hands that have exactly two diamonds.

**Hint: DON'T use the random_state parameter in this code.**

In [55]:
twodiamonds_count = 0
n_iterations = 10000

for i in range(n_iterations):
    new_hand = cards.sample(5)
    twodiamonds_count += hastwodiamonds(new_hand)

In [59]:
print('twodiamonds_count =', twodiamonds_count)
print('n_iterations =', n_iterations)

print('Proportion of hands (out of 10000) that contain exactly two diamonds is equal to = ', (twodiamonds_count/n_iterations))

twodiamonds_count = 2702
n_iterations = 10000
Proportion of hands (out of 10000) that contain exactly two diamonds is equal to =  0.2702


## Question 3:  [1 pts] Compare your approximated probability of drawing a hand with exactly two diamonds (from question 2) and the probability that you calculated (in question 1). What is the difference?

The probability of drawing a hand with exactly two diamonds that I calculated in question 1 was .2743, and the probability in question two is equals to .2702. The difference between these two numbers is .2743 - .2702 = 0.0041 = 0.41%

## Question 4: Sampling Cards with Replacement

### 4a:  [1 pts] Randomly sample 3 cards from the cards dataframe with replacement.

In [75]:
cards.sample(3, replace = True)

color   suit face
13  black  spade    A
13  black  spade    A
44    red  heart    6

### 4b.  [2 pts] Assume that we randomly sampled 3 cards from the deck with replacement. How many three card permutations exist for this experiment?

#### Hint: "Permutation" imples we care about the order in which the cards were drawn.

The way to calculate the number of replacements with permutations is to take the number of cards in the deck (52) and put that to the power of the number of cards we are sampling (3)

This results in 52^(3) = 140608

In [58]:
print('140608')

140608
